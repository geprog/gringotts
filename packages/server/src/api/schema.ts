import { FastifyInstance } from 'fastify';

export function addSchemas(server: FastifyInstance): void {
  server.addSchema({
    $id: 'SuccessResponse',
    type: 'object',
    description: 'Success response',
    properties: {
      ok: { type: 'boolean' },
    },
  });

  server.addSchema({
    $id: 'ErrorResponse',
    type: 'object',
    description: 'Error response',
    properties: {
      error: { type: 'string' },
    },
  });

  server.addSchema({
    $id: 'ProjectInvoiceData',
    type: 'object',
    properties: {
      _id: { type: 'string' },
      email: { type: 'string' },
      name: { type: 'string' },
      addressLine1: { type: 'string' },
      addressLine2: { type: 'string' },
      zipCode: { type: 'string' },
      city: { type: 'string' },
      country: { type: 'string' },
    },
  });

  server.addSchema({
    $id: 'Project',
    type: 'object',
    properties: {
      _id: { type: 'string' },
      name: { type: 'string' },
      apiToken: { type: 'string' },
      mollieApiKey: { type: 'string' },
      paymentProvider: { type: 'string' },
      webhookUrl: { type: 'string' },
      invoiceData: { $ref: 'ProjectInvoiceData' },
      currency: { type: 'string' },
      vatRate: { type: 'number' },
    },
  });

  server.addSchema({
    $id: 'Payment',
    type: 'object',
    properties: {
      _id: { type: 'string' },
      type: { type: 'string' },
      status: { type: 'string' },
      currency: { type: 'string' },
      amount: { type: 'number' },
      description: { type: 'number' },
    },
  });

  server.addSchema({
    $id: 'InvoiceItem',
    type: 'object',
    properties: {
      _id: { type: 'string' },
      description: { type: 'string' },
      units: { type: 'number' },
      pricePerUnit: { type: 'number' },
    },
  });

  server.addSchema({
    $id: 'Invoice',
    type: 'object',
    properties: {
      _id: { type: 'string' },
      date: { type: 'string' },
      sequentialId: { type: 'number' },
      items: {
        type: 'array',
        items: {
          $ref: 'InvoiceItem',
        },
      },
      subscription: { type: 'object', properties: { _id: { type: 'string' } }, additionalProperties: false },
      // subscription: { $ref: 'Subscription' },
      customer: { $ref: 'Customer' },
      status: { type: 'string' },
      currency: { type: 'string' },
      vatRate: { type: 'number' },
      amount: { type: 'number' },
      vatAmount: { type: 'number' },
      totalAmount: { type: 'number' },
      number: { type: 'string' },
    },
  });

  server.addSchema({
    $id: 'Customer',
    type: 'object',
    properties: {
      _id: { type: 'string' },
      email: { type: 'string' },
      name: { type: 'string' },
      addressLine1: { type: 'string' },
      addressLine2: { type: 'string' },
      zipCode: { type: 'string' },
      city: { type: 'string' },
      country: { type: 'string' },
      invoicePrefix: { type: 'string' },
      invoiceCounter: { type: 'string' },
      balance: { type: 'number' },
      activePaymentMethod: { $ref: 'PaymentMethod' },
    },
  });

  server.addSchema({
    $id: 'SubscriptionChange',
    type: 'object',
    properties: {
      _id: { type: 'string' },
      start: { type: 'string' },
      end: { type: 'string' },
      pricePerUnit: { type: 'number' },
      units: { type: 'number' },
    },
  });

  server.addSchema({
    $id: 'Subscription',
    type: 'object',
    properties: {
      _id: { type: 'string' },
      anchorDate: { type: 'string' },
      status: { type: 'string' },
      error: { type: 'string' },
      lastPayment: { type: 'string' },
      currentPeriodStart: { type: 'string' },
      currentPeriodEnd: { type: 'string' },
      activeUntil: { type: 'string' },
      customer: { $ref: 'Customer' },
      changes: {
        type: 'array',
        items: { $ref: 'SubscriptionChange' },
      },
      // invoices: {
      //   type: 'array',
      //   items: { $ref: 'Invoice' },
      // },
    },
  });

  server.addSchema({
    $id: 'PaymentMethod',
    type: 'object',
    properties: {
      _id: { type: 'string' },
      customer: { $ref: 'Customer' },
      type: { type: 'string' },
      name: { type: 'string' },
    },
  });
}
