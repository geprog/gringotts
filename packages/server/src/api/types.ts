import { Project } from '~/entities';

declare module 'fastify' {
  export interface FastifyRequest {
    project?: Project;
    admin?: boolean;
  }
}
