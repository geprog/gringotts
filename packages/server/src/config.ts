import dotenv from 'dotenv';
import path from 'path';

dotenv.config({
  path: path.join(__dirname, '..', '..', '..', '.env'),
});

const defaultPort = 3000;

export const config = {
  port: process.env.PORT ? parseInt(process.env.PORT) : defaultPort,
  publicUrl: process.env.PUBLIC_URL || `http://localhost:${defaultPort}`,
  postgresUrl: process.env.POSTGRES_URL as string,
  adminToken: process.env.ADMIN_TOKEN as string,
  jwtSecret: process.env.JWT_SECRET as string,
  gotenbergUrl: process.env.GOTENBERG_URL || 'http://localhost:3000',
  dataPath: process.env.STORAGE_PATH || path.join(__dirname, 'data'),
};

export function checkConfig(): void {
  const _config = config as Partial<typeof config>;

  if (!_config.publicUrl) {
    throw new Error('Please configure PUBLIC_URL');
  }

  if (!_config.postgresUrl) {
    throw new Error('Please configure POSTGRES_URL');
  }

  if (!_config.adminToken) {
    throw new Error('Please configure ADMIN_TOKEN');
  }
}
