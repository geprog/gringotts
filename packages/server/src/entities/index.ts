import { Customer, customerSchema } from './customer';
import { Subscription, subscriptionSchema } from './subscription';
import { SubscriptionChange, subscriptionChangeSchema } from './subscription_change';

export { Customer, customerSchema, Subscription, SubscriptionChange, subscriptionChangeSchema, subscriptionSchema };
