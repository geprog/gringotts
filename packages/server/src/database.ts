import { MikroORM } from '@mikro-orm/core';
import { EntityManager, EntityRepository, PostgreSqlDriver } from '@mikro-orm/postgresql';

import { config } from '~/config';
import {
  Customer,
  customerSchema,
  Invoice,
  InvoiceItem,
  invoiceItemSchema,
  invoiceSchema,
  Payment,
  paymentSchema,
  Subscription,
  SubscriptionChange,
  subscriptionChangeSchema,
  subscriptionSchema,
} from '~/entities';
import { addExitHook } from '~/lib/exit_hooks';

export class Database {
  orm!: MikroORM;

  async init(): Promise<void> {
    if (!config.postgresUrl) {
      throw new Error('POSTGRES_URL is not set');
    }

    this.orm = await MikroORM.init<PostgreSqlDriver>({
      type: 'postgresql',
      clientUrl: config.postgresUrl,
      entities: [
        customerSchema,
        subscriptionSchema,
        subscriptionChangeSchema,
        paymentSchema,
        invoiceSchema,
        invoiceItemSchema,
      ],
      discovery: { disableDynamicFileAccess: true },
    });
  }

  async connect(): Promise<void> {
    await this.orm.connect();

    const generator = this.orm.getSchemaGenerator();
    await generator.updateSchema();

    addExitHook(() => this.orm.close());
  }

  get em(): EntityManager {
    return this.orm.em.fork() as EntityManager;
  }

  get subscriptions(): EntityRepository<Subscription> {
    return this.em.getRepository(Subscription);
  }

  get customers(): EntityRepository<Customer> {
    return this.em.getRepository(Customer);
  }

  get subscriptionChanges(): EntityRepository<SubscriptionChange> {
    return this.em.getRepository(SubscriptionChange);
  }

  get payments(): EntityRepository<Payment> {
    return this.em.getRepository(Payment);
  }

  get invoices(): EntityRepository<Invoice> {
    return this.em.getRepository(Invoice);
  }

  get invoiceItems(): EntityRepository<InvoiceItem> {
    return this.em.getRepository(InvoiceItem);
  }
}

export const database = new Database();
