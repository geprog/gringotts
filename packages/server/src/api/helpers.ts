import { FastifyRequest } from 'fastify';

import { Project } from '~/entities';

// eslint-disable-next-line @typescript-eslint/require-await
export async function getProjectFromRequest(request: FastifyRequest): Promise<Project> {
  if (!request.project) {
    throw new Error('request.project should be defined');
  }

  return request.project;
}
