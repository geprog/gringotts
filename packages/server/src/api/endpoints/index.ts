import { FastifyInstance } from 'fastify';

import { config } from '~/config';
import { database } from '~/database';

import { customerEndpoints } from './customer';
import { invoiceEndpoints } from './invoice';
import { mockedCheckoutEndpoints } from './mocked_checkout';
import { paymentEndpoints } from './payment';
import { paymentMethodEndpoints } from './payment_method';
import { projectEndpoints } from './project';
import { subscriptionEndpoints } from './subscription';

export async function apiEndpoints(server: FastifyInstance): Promise<void> {
  server.addHook('onRequest', async (request, reply) => {
    // skip nuxt api routes
    if (request.url === '/api/auth/login' || request.url === '/api/auth/logout' || request.url === '/api/user') {
      return;
    }

    if (request.routerPath === '/api/invoice/download') {
      return;
    }

    if (request.routerPath === '/api/payment/webhook/:projectId') {
      return;
    }

    if (request.routerPath === '/api/mocked/checkout/:paymentId') {
      return;
    }

    const apiToken =
      (request.headers?.authorization || '').replace('Bearer ', '') || (request.query as { token: string }).token;
    if (!apiToken) {
      await reply.code(401).send({ error: 'Missing api token' });
      return reply;
    }

    if (request.routerPath?.startsWith('/api/project') && request.url !== '/api/project/token-project') {
      if (apiToken === config.adminToken) {
        request.admin = true;
        return;
      }

      await reply.code(401).send({ error: 'You need to have admin access' });
      return reply;
    }

    const project = await database.projects.findOne({ apiToken }, { populate: ['invoiceData'] });
    if (!project) {
      await reply.code(401).send({ error: 'Invalid api token' });
      return reply;
    }

    request.project = project;
  });

  await server.register(subscriptionEndpoints);
  await server.register(customerEndpoints);
  await server.register(invoiceEndpoints);
  await server.register(paymentEndpoints);
  await server.register(projectEndpoints);
  await server.register(paymentMethodEndpoints);
  await server.register(mockedCheckoutEndpoints);
}
