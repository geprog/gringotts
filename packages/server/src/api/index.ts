import fastifyFormBody from '@fastify/formbody';
import fastifyHelmet from '@fastify/helmet';
import fastifyStatic from '@fastify/static';
import fastifySwagger from '@fastify/swagger';
import fastifyView from '@fastify/view';
import fastify, { FastifyInstance } from 'fastify';
import Handlebars from 'handlebars';
import path from 'path';

import { config } from '~/config';
import { database } from '~/database';
import { Invoice } from '~/entities';
import { Currency } from '~/entities/payment';
import { formatDate } from '~/lib/dayjs';

import { customerEndpoints } from './endpoints/customer';
import { invoiceEndpoints } from './endpoints/invoice';
import { paymentEndpoints } from './endpoints/payment';
import { projectEndpoints } from './endpoints/project';
import { subscriptionEndpoints } from './endpoints/subscription';
import { addSchemas } from './schema';

export async function init(): Promise<FastifyInstance> {
  const server = fastify({
    logger: {
      transport:
        process.env.NODE_ENV === 'production'
          ? undefined
          : {
              target: 'pino-pretty',
              options: {
                translateTime: 'HH:MM:ss Z',
                ignore: 'pid,hostname',
              },
            },
    },
  });

  server.addHook('onRequest', async (request, reply) => {
    if (!request.routerPath) {
      await reply.code(404).send({
        error: 'Not found',
      });
      return reply;
    }

    // skip requests to our docs
    if (request.routerPath?.startsWith('/documentation')) {
      return;
    }

    if (request.routerPath?.startsWith('/static')) {
      return;
    }

    if (request.routerPath === '/payment/webhook/:projectId') {
      return;
    }

    const apiToken =
      (request.headers?.authorization || '').replace('Bearer ', '') || (request.query as { token: string }).token;
    if (!apiToken) {
      await reply.code(401).send({ error: 'Missing api token' });
      return reply;
    }

    if (request.routerPath?.startsWith('/project')) {
      if (apiToken === config.adminToken) {
        request.admin = true;
        return;
      }

      await reply.code(401).send({ error: 'You need to have admin access' });
      return reply;
    }

    const project = await database.projects.findOne({ apiToken }, { populate: ['invoiceData'] });
    if (!project) {
      await reply.code(401).send({ error: 'Invalid api token' });
      return reply;
    }

    request.project = project;
  });

  await server.register(fastifyFormBody);

  await server.register(fastifyHelmet);

  await server.register(fastifyView, {
    engine: {
      handlebars: Handlebars,
    },
  });

  await server.register(fastifyStatic, {
    root:
      process.env.NODE_ENV === 'production'
        ? path.join(__dirname, 'public')
        : path.join(__dirname, '..', '..', 'public'),
    prefix: '/static',
  });

  Handlebars.registerHelper('formatDate', (date: Date, format: string) => formatDate(date, format));
  Handlebars.registerHelper('amountToPrice', (amount: number, currency: Currency) =>
    Invoice.amountToPrice(amount, currency),
  );

  await server.register(fastifySwagger, {
    routePrefix: '/documentation',
    swagger: {
      info: {
        title: 'Gringotts api',
        description: 'Documentation for the Gringotts api',
        version: '0.1.0',
      },
      host: 'localhost:3000',
      schemes: ['http', 'https'],
      consumes: ['application/json'],
      produces: ['application/json'],
      securityDefinitions: {
        authorization: {
          type: 'apiKey',
          name: 'Authorization',
          in: 'header',
        },
      },
      security: [{ authorization: [] }],
    },
    refResolver: {
      buildLocalReference(json) {
        if (!json.title && json.$id) {
          json.title = json.$id;
        }
        return json.$id as string;
      },
    },
    exposeRoute: true,
  });

  addSchemas(server);

  subscriptionEndpoints(server);
  customerEndpoints(server);
  invoiceEndpoints(server);
  paymentEndpoints(server);
  projectEndpoints(server);

  return server;
}
