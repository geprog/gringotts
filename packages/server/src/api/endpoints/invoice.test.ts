import { beforeAll, describe, expect, it, vi } from 'vitest';

import { getFixtures } from '~/../test/fixtures';
import { init as apiInit } from '~/api';
import * as config from '~/config';
import * as database from '~/database';
import { Invoice } from '~/entities';

describe('Invoice endpoints', () => {
  beforeAll(async () => {
    vi.spyOn(config, 'config', 'get').mockReturnValue(getFixtures().config);

    await database.database.init();
  });

  it('should get an invoice', async () => {
    // given
    const testData = getFixtures();

    vi.spyOn(database, 'database', 'get').mockReturnValue({
      invoices: {
        findOne() {
          return Promise.resolve(testData.invoice);
        },
      },
      customers: {
        findOne() {
          return Promise.resolve(testData.customer);
        },
      },
      subscriptions: {
        findOne() {
          return Promise.resolve(testData.subscription);
        },
      },
      projects: {
        findOne() {
          return Promise.resolve(testData.project);
        },
      },
    } as unknown as database.Database);

    const server = await apiInit();

    // when
    const response = await server.inject({
      method: 'GET',
      headers: {
        authorization: `Bearer ${testData.project.apiToken}`,
      },
      url: `/api/invoice/${testData.invoice._id}`,
    });

    // then
    expect(response.statusCode).toBe(200);

    const responseData: Invoice = response.json();
    expect(responseData._id).toBe(testData.invoice._id);
    expect(responseData.amount).toBe(214.74);
    expect(responseData.totalAmount).toBe(Invoice.roundPrice(214.74 * 1.19));
  });

  it('should get an html invoice', async () => {
    // given
    const testData = getFixtures();

    vi.spyOn(database, 'database', 'get').mockReturnValue({
      invoices: {
        findOne() {
          return Promise.resolve(testData.invoice);
        },
      },
      customers: {
        findOne() {
          return Promise.resolve(testData.customer);
        },
      },
      subscriptions: {
        findOne() {
          return Promise.resolve(testData.subscription);
        },
      },
      projects: {
        findOne() {
          return Promise.resolve(testData.project);
        },
      },
    } as unknown as database.Database);

    const server = await apiInit();

    // when
    const response = await server.inject({
      method: 'GET',
      headers: {
        authorization: `Bearer ${testData.project.apiToken}`,
      },
      url: `/api/invoice/${testData.invoice._id}/html`,
    });

    // then
    expect(response.statusCode).toBe(200);

    const responseData = response.body;
    expect(responseData).toMatchSnapshot();
  });
});
