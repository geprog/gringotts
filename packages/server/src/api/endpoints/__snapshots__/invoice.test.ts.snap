// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Invoice endpoints > should get an html invoice 1`] = `
"<html>

<!-- Based on: https://github.com/sparksuite/simple-html-invoice-template -->

<head>
  <meta charset='utf-8' />
  <title>Invoice</title>

  <link rel=\\"stylesheet\\" href=\\"/static/styles/invoice.css\\" />
</head>

<body>
  <div class='invoice-box'>
    <table cellpadding='0' cellspacing='0'>
      <tr class='top'>
        <td colspan='3'>
          <table>
            <tr>
              <td class='title'>
                <img src='http://localhost/my-logo.png' style='width: 100%; max-width: 150px' />
              </td>

              <td>
                <div style=\\"text-align: left; width: fit-content; margin-left: auto\\">
                  <table>
                    <tr>
                      <td>Number:</td>
                      <td>INV-F1B-0B6H-002</td>
                    </tr>
                    <tr>
                      <td>Date:</td>
                      <td>01.02.2020</td>
                    </tr>
                  </table>
                </div>
              </td>
            </tr>
          </table>
        </td>
      </tr>

      <tr class='information'>
        <td colspan='3'>
          <table>
            <tr>
              <td>
                John Doe<br />
                BigBen Street 954<br />
                123<br />
                ENG-1234 London<br />
                GB
              </td>

              <td>
                <div style=\\"text-align: left; width: fit-content; margin-left: auto\\">
                  Test company<br />
                  My Street 123<br />
                  Postbox 321<br />
                  GB-12345 London<br />
                  GB<br />
                  test@example.tld
                </div>
              </td>
            </tr>
          </table>
        </td>
      </tr>

      <tr class='heading'>
        <td>Item</td>
        <td>Units</td>
        <td>Price</td>
      </tr>

      <tr class=\\"item\\">
        <td style=\\"white-space: pre-wrap;\\">Test item</td>
        <td>13</td>
        <td>12.34 €</td>
      </tr>
      <tr class=\\"item last\\">
        <td style=\\"white-space: pre-wrap;\\">Second test item</td>
        <td>1</td>
        <td>54.32 €</td>
      </tr>

      <tr class=\\"sum\\">
        <td></td>

        <td colspan=\\"2\\">
          <table>
            <tr>
              <td>Amount</td>
              <td>214.74 €</td>
            </tr>
            <tr>
              <td>Vat (19%)</td>
              <td>40.80 €</td>
            </tr>
            <tr class=\\"total\\">
              <td>Total</td>
              <td>255.54 €</td>
            </tr>
          </table>
        </td>
      </tr>
    </table>

    <p>The total amount will be debited from your bank account soon.</p>
  </div>
</body>

</html>"
`;
